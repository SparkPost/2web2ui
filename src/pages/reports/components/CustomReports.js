import React from 'react';
import { connect } from 'react-redux';
import { addFilters, clearFilters, refreshReportOptions } from 'src/actions/reportOptions';
import { selectCustomReports } from 'src/selectors/customReports';
import { Save } from '@sparkpost/matchbox-icons';
import { saveReport, deleteReport } from 'src/actions/customReports';
import { withRouter } from 'react-router-dom';
import { Button, Panel, TextField } from 'src/components/matchbox';
import CustomReportsList from './CustomReportsList';
import { parseSearch } from 'src/helpers/reports';
import qs from 'query-string';
import _ from 'lodash';

import styles from './CustomReports.module.scss';

const REPORT_LIMIT = 10;

export function CustomReports(props) {
  const {
    addFilters,
    deleteReport,
    clearFilters,
    searchOptions,
    loading,
    refreshReportOptions,
    reports,
    saveReport,
  } = props;

  const [name, setName] = React.useState('');

  function handleNameChange(e) {
    const value = e.target.value;

    // Limit string length
    if (value.length < 24) {
      setName(e.target.value);
    }
  }

  function handleSave() {
    let toSave = searchOptions;

    if (searchOptions.range !== 'custom') {
      toSave = _.omit(toSave, ['to', 'from']);
    }

    saveReport({
      name: name.trim(),
      url: qs.stringify(toSave),
    });
  }

  function handleLoad(report) {
    const { options, filters = [] } = parseSearch(report.url);
    clearFilters();
    addFilters(filters);
    refreshReportOptions(options);
    setName(report.name);
  }

  const isLimited = reports.length >= REPORT_LIMIT;
  const canOverwrite = _.find(reports, ['name', name.trim()]);
  const saveDisabled = name === '' || loading;

  return (
    <Panel.LEGACY.Section>
      <div className={styles.Wrap}>
        <div className={styles.NameWrapper}>
          <TextField
            id="custom-report-name"
            label="Report Name"
            labelHidden
            placeholder="Untitled Report"
            onChange={handleNameChange}
            value={name}
          />
        </div>

        {!isLimited || (canOverwrite && isLimited) ? (
          <Button disabled={saveDisabled} onClick={handleSave} flat color="orange">
            <Save className={styles.SaveIcon} />
            {loading ? 'Saving...' : 'Save'}
          </Button>
        ) : null}

        {isLimited && (
          <div className={styles.Limited}>You are limited to {REPORT_LIMIT} saved reports</div>
        )}

        <div className={styles.ReportsButton}>
          <CustomReportsList
            deleting={loading}
            handleLoad={handleLoad}
            reports={reports}
            onDelete={deleteReport}
          />
        </div>
      </div>
    </Panel.LEGACY.Section>
  );
}

const mapStateToProps = state => ({
  reports: selectCustomReports(state),
  loading: state.currentUser.userOptionsPending,
});

const mapDispatchToProps = {
  addFilters,
  clearFilters,
  refreshReportOptions,
  saveReport,
  deleteReport,
};

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(CustomReports));
