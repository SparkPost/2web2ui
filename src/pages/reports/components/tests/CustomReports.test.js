import React from 'react';
import { shallow } from 'enzyme';
import { CustomReports } from '../CustomReports';

describe('Component: CustomReports', () => {
  let wrapper;
  let props;

  beforeEach(() => {
    props = {
      loading: false,
      reports: [
        { name: 'report 1', url: 'report1url' },
        {
          name: 'report 2',
          url: 'range=custom&metrics=count_delayed&filters=Sending%20Domain%3Atest.com'
        }
      ],
      addFilters: jest.fn(),
      clearFilters: jest.fn(),
      refreshReportOptions: jest.fn(),
      saveReport: jest.fn(),
      deleteReport: jest.fn(),
      searchOptions: { range: 'day', to: 'to', from: 'from' }
    };
    wrapper = shallow(<CustomReports {...props} />);
  });

  it('should render', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('report name field', () => {
    it('should handle input change', () => {
      wrapper.find('TextField').simulate('change', { target: { value: 'test-name ' }});
      expect(wrapper.find('TextField').prop('value')).toEqual('test-name ');
    });

    it('should enabled the save button with a value', () => {
      wrapper.find('TextField').simulate('change', { target: { value: 'test-name ' }});
      expect(wrapper.find('Button').prop('disabled')).toBe(false);
    });
    it('should limit input to 24 characters', () => {
      wrapper.find('TextField').simulate('change', { target: { value: 'test-name' }});
      expect(wrapper.find('TextField').prop('value')).toEqual('test-name');
      wrapper.find('TextField').simulate('change', { target: {
        value: 'a-very-very-very-long-test-name'
      }});
      expect(wrapper.find('TextField').prop('value')).toEqual('test-name');
    });
  });

  describe('save button', () => {
    it('should handle saving a report', () => {
      wrapper.find('TextField').simulate('change', { target: { value: 'test-name ' }});
      expect(wrapper.find('Button').prop('disabled')).toBe(false);
      wrapper.find('Button').simulate('click');
      expect(props.saveReport).toHaveBeenCalledWith({
        name: 'test-name',
        url: 'range=day'
      });
    });

    it('should handle saving a report with custom dates', () => {
      wrapper.setProps({ searchOptions: { ...props.searchOptions, range: 'custom' }});
      wrapper.find('TextField').simulate('change', { target: { value: 'test-name ' }});
      wrapper.find('Button').simulate('click');
      expect(props.saveReport).toHaveBeenCalledWith({
        name: 'test-name',
        url: 'from=from&range=custom&to=to'
      });
    });

    it('should be disabled when saving a report', () => {
      wrapper.setProps({ loading: true });
      expect(wrapper.find('Button').prop('disabled')).toBe(true);
    });

    it('should not be visible when exceeding report limit', () => {
      wrapper.setProps({ reports: new Array(10) });
      expect(wrapper.find('Button')).not.toExist();
    });

    it('should be visible when exceeding report limit, but able to overwrite an existing report', () => {
      const newReports = new Array(10);
      newReports.push(props.reports[0]);

      wrapper.setProps({ reports: newReports });
      wrapper.find('TextField').simulate('change', { target: { value: 'report 1 ' }});
      expect(wrapper.find('Button')).toExist();
    });
  });

  it('should show a message when exceeding report limit', () => {
    wrapper.setProps({ reports: new Array(11) });
    expect(wrapper.find('.Limited').text()).toEqual('You are limited to 10 saved reports');
  });

  it('should correctly load a report', () => {
    wrapper.find('withRouter(CustomReportsList)').prop('handleLoad')(props.reports[1]);
    expect(wrapper.find('TextField').prop('value')).toEqual('report 2');
    expect(props.clearFilters).toHaveBeenCalled();

    expect(props.addFilters).toHaveBeenCalledWith([{
      id: undefined,
      type: 'Sending Domain',
      value: 'test.com'
    }]);

    expect(props.refreshReportOptions).toHaveBeenCalledWith({
      metrics: ['count_delayed'],
      relativeRange: 'custom'
    });
  });
});
