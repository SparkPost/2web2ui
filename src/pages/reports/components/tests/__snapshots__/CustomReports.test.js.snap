// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: CustomReports should render 1`] = `
<Panel.LEGACY.Section>
  <div
    className="Wrap"
  >
    <div
      className="NameWrapper"
    >
      <TextField
        id="custom-report-name"
        label="Report Name"
        labelHidden={true}
        onChange={[Function]}
        placeholder="Untitled Report"
        value=""
      />
    </div>
    <Button
      color="orange"
      disabled={true}
      flat={true}
      onClick={[Function]}
    >
      <Save
        className="SaveIcon"
      />
      Save
    </Button>
    <div
      className="ReportsButton"
    >
      <withRouter(CustomReportsList)
        deleting={false}
        handleLoad={[Function]}
        onDelete={[MockFunction]}
        reports={
          Array [
            Object {
              "name": "report 1",
              "url": "report1url",
            },
            Object {
              "name": "report 2",
              "url": "range=custom&metrics=count_delayed&filters=Sending%20Domain%3Atest.com",
            },
          ]
        }
      />
    </div>
  </div>
</Panel.LEGACY.Section>
`;
